File: BreakOut3b/BOSettingsTableViewController.m
//
//  BOSettingsTableViewController.m
//  BreakOut3b
//
//  Created by Carl Tengwall on 2012-11-09.
//  Copyright (c) 2012 Carl Tengwall. All rights reserved.
//

#import "BOSettingsTableViewController.h"

@interface BOSettingsTableViewController ()

@end

@implementation BOSettingsTableViewController

@synthesize boNewUserViewController = _boNewUserViewController;
@synthesize users = _users;
@synthesize player = _player;


- (void)setUsers:(NSMutableArray *)users {
    if (!users) NSLog(@"Who is trying to kill users?");
    else _users = users;
}

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    //_ NSLog(@"viewDidLoad");
    [super viewDidLoad];

    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
 
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    self.navigationItem.leftBarButtonItem = self.editButtonItem;

    
    int noOfUsers = [[NSUserDefaults standardUserDefaults] integerForKey:@"noOfUsers"];
    int currentPlayer = [[NSUserDefaults standardUserDefaults] integerForKey:@"currentPlayer"];

    
    //_ NSLog(@"In BOSettingsTableViewController viewDidLoad (1) with noOfUsers=%d and currentPlayer=%d", noOfUsers, currentPlayer);
    if(currentPlayer==-1)
    {
        NSLog(@"This is a cold start");

        [[NSUserDefaults standardUserDefaults] setInteger:0 forKey:@"noOfUsers"];
         noOfUsers = [[NSUserDefaults standardUserDefaults] integerForKey:@"noOfUsers"];
        //_ NSLog(@"In BOSettingsTableViewController viewDidLoad (2) with noOfUsers=%d", noOfUsers);
        [[NSUserDefaults standardUserDefaults] synchronize];
    }
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [[self tableView] reloadData];
    
  //  [self.userIdTextField becomeFirstResponder];
  //  self.userIdTextField.delegate = self;
}
- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    //_ NSLog(@"In BOSettingsTableViewController prepareForSegue:");
    if ([[segue identifier] isEqualToString:@"pushPlay"]) {
         //_ NSLog(@"pushPlay");
         BOViewController *bovc = segue.destinationViewController;
         bovc.users = self.users;
         //Send the user to the Play view
         NSIndexPath *selectedRowIndex = [self.tableView indexPathForSelectedRow];
        
        // Set the selected user 
        [[NSUserDefaults standardUserDefaults] setInteger:selectedRowIndex.row forKey:@"currentPlayer"];
        [[NSUserDefaults standardUserDefaults] synchronize];
        
    }
    if ([segue.identifier isEqualToString:@"linkModal"]) {
        //_ NSLog(@"linkModal");
        UINavigationController *navc = segue.destinationViewController;
        
        BONewUserViewController *bonuvc = (BONewUserViewController *)navc.topViewController;
        bonuvc.delegate = self;
        //_ NSLog(@"bonuvc.delegate = %@", bonuvc.delegate);
    }
}
#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    //_ NSLog(@"In BOSettingsTableViewController numberOfSectionsInTableView:");
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    //_ NSLog(@"In BOSettingsTableViewController numberOfRowsInSection:");
    // Return the number of rows in the section.
    return self.users.count;
}
/*
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    // Configure the cell...
    
    return cell;
}*/

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    //_ NSLog(@"In BOSettingsTableViewController cellForRowAtIndexPath:");
    // Hämta cell
    static NSString *reuseIdentifier = @"BOUserCell";
    BOUserCell *cell = [tableView dequeueReusableCellWithIdentifier:reuseIdentifier];
    
    if (!cell) // Lazy allocation
        cell = [[BOUserCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"BOUserCell"];
    
    //_ NSLog(@"cell = %@",cell);
    
    // Hämta data
    BOUser *user = [self.users objectAtIndex:indexPath.row];
    //_ NSLog(@"user = %@, uid = %@, difficulty = %@, score = %@, background = %@",user,user.userId,user.difficulty,user.score, user.backgroundColor);
    
    // Konfigurera cellen
    cell.userIdLabel.text = user.userId;
    cell.scoreLabel.text = [NSString stringWithFormat:@"%@", user.score];
    return cell;
}
/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

- (void)tableView:(UITableView *)tableView commitEditingStyle:
(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    //_ NSLog(@"commitEditingStyle:");
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        
        //[self.links removeObjectAtIndex:indexPath.row];
        NSMutableArray *su = [[NSMutableArray alloc] initWithArray:self.users];
        [su removeObjectAtIndex:indexPath.row];
        self.users = su;
        
        int noOfUsers = [[NSUserDefaults standardUserDefaults] integerForKey:@"noOfUsers"];
        int currentPlayer = [[NSUserDefaults standardUserDefaults] integerForKey:@"currentPlayer"];
        NSArray *oldUsers = [[NSUserDefaults standardUserDefaults] objectForKey:@"users"];
        NSArray *oldDiffs = [[NSUserDefaults standardUserDefaults] objectForKey:@"diffs"];
        NSArray *oldScores = [[NSUserDefaults standardUserDefaults] objectForKey:@"scores"];
        NSArray *oldColors = [[NSUserDefaults standardUserDefaults] objectForKey:@"colors"];
        
        NSMutableArray *tempUsers = [[NSMutableArray alloc] initWithArray:oldUsers];
        NSMutableArray *tempDiffs = [[NSMutableArray alloc] initWithArray:oldDiffs];
        NSMutableArray *tempScores = [[NSMutableArray alloc] initWithArray:oldScores];
        NSMutableArray *tempColors = [[NSMutableArray alloc] initWithArray:oldColors];
        
        
        [tempUsers removeObjectAtIndex:indexPath.row];
        [tempDiffs removeObjectAtIndex:indexPath.row];
        [tempScores removeObjectAtIndex:indexPath.row];
        [tempColors removeObjectAtIndex:indexPath.row];
        
        [[NSUserDefaults standardUserDefaults] setObject:tempUsers forKey:@"users"];
        [[NSUserDefaults standardUserDefaults] setObject:tempDiffs forKey:@"diffs"];
        [[NSUserDefaults standardUserDefaults] setObject:tempScores forKey:@"scores"];
        [[NSUserDefaults standardUserDefaults] setObject:tempColors forKey:@"colors"];
        
        [[NSUserDefaults standardUserDefaults] setInteger:noOfUsers-1 forKey:@"noOfUsers"];
        if (currentPlayer == indexPath.row) {
            currentPlayer = 0;
        } else if (currentPlayer > indexPath.row) {
            currentPlayer = currentPlayer - 1;
        }
        [[NSUserDefaults standardUserDefaults] setInteger:currentPlayer forKey:@"currentPlayer"];
        
        [[NSUserDefaults standardUserDefaults] synchronize];
        
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                         withRowAnimation:UITableViewRowAnimationAutomatic];
    }
    else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }
}

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    //_ NSLog(@"didSelectRowAtIndexPath:");
    // Push in in BOViewController
    //[self performSegueWithIdentifier:@"pushPlay" sender:self];

}
- (void)boNewUserViewController:(BONewUserViewController *) sender
                      gotUserId:(NSString *) uid
                  gotDifficulty:(NSNumber *) diff
                          score:(NSNumber *) score
          andGotBackgroundColor:(NSNumber *) bkg 
{
    //_ NSLog(@"In BOSettingsTableViewController boNewUserViewController:");

    NSMutableArray *userArray = [[NSMutableArray alloc] init];
    
    BOUser *newUser = [[BOUser alloc] initWithId:uid difficulty:diff score: score
                           andBackground:bkg];
    [self.users arrayByAddingObject:newUser];
    
    // Save new user for the future
    if (newUser) {
        //_ NSLog(@"newUser");
        NSArray *newUsers = [[[NSUserDefaults standardUserDefaults] objectForKey:@"users"] arrayByAddingObject:uid];
        NSArray *newDiffs = [[[NSUserDefaults standardUserDefaults] objectForKey:@"diffs"] arrayByAddingObject:diff];
        NSArray *newScores = [[[NSUserDefaults standardUserDefaults] objectForKey:@"scores"] arrayByAddingObject:[NSNumber numberWithInt:0]];
        NSArray *newColors = [[[NSUserDefaults standardUserDefaults] objectForKey:@"colors"] arrayByAddingObject:bkg];
        
        int noOfUsers = 1 + [[NSUserDefaults standardUserDefaults] integerForKey:@"noOfUsers"];
    
        NSDictionary *appSettings = [NSDictionary dictionaryWithObjectsAndKeys:
                                       [NSNumber numberWithInt:noOfUsers-1], @"currentPlayer"
                                     , [NSNumber numberWithInt:noOfUsers], @"noOfUsers"
                                     , newUsers , @"users"
                                     , newDiffs , @"diffs"
                                     , newScores , @"scores"
                                     , newColors, @"colors"
                                     , nil];
        // Register the new settings
        [[NSUserDefaults standardUserDefaults] registerDefaults:appSettings];
        [[NSUserDefaults standardUserDefaults] synchronize];
        
        for (NSInteger i=0; i<newUsers.count;i++) {
            
            NSString *myUid = (NSString *)[newUsers objectAtIndex:i];
            NSNumber *myDiff = (NSNumber *)[newDiffs objectAtIndex:i];
            NSNumber *myScore = (NSNumber *)[newScores objectAtIndex:i];
            NSNumber *myColor = (NSNumber *)[newColors objectAtIndex:i];
            
        
            BOUser *aUser = [[BOUser alloc] initWithId:myUid difficulty:myDiff score: myScore
                                           andBackground:myColor];
            [userArray addObject:aUser];
        }
        
        self.users = [NSArray arrayWithArray:userArray];
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
    [[self tableView] reloadData];
}

- (void)newScore:(BOViewController *) sender
       withScore: (int) myScore {
       self.users = sender.users;
    NSLog(@"Score=%d",myScore);
    [self dismissViewControllerAnimated:YES completion:nil];
    [[self tableView] reloadData];

    
}

@end

File: BreakOut3b/BOSettingsTableViewController.swift
import UIKit

@objc(BOSettingsTableViewController)
final class BOSettingsTableViewController: UITableViewController, @MainActor BONewUserViewControllerDelegate, @MainActor BOViewControllerDelegate {
    @objc var boNewUserViewController: BONewUserViewController?
    @objc var users: NSMutableArray = []
    @objc var player: BOUser?

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.leftBarButtonItem = editButtonItem

        let defaults = UserDefaults.standard
        let currentPlayer = defaults.integer(forKey: "currentPlayer")
        if currentPlayer == -1 {
            NSLog("This is a cold start")
            defaults.set(0, forKey: "noOfUsers")
            defaults.synchronize()
        }
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        tableView.reloadData()
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "pushPlay" {
            if let destination = segue.destination as? BOViewController,
               let selectedRowIndex = tableView.indexPathForSelectedRow {
                destination.users = users
                UserDefaults.standard.set(selectedRowIndex.row, forKey: "currentPlayer")
                UserDefaults.standard.synchronize()
            }
        }

        if segue.identifier == "linkModal",
           let navigationController = segue.destination as? UINavigationController,
           let newUserViewController = navigationController.topViewController as? BONewUserViewController {
            newUserViewController.delegate = self
        }
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        users.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let reuseIdentifier = "BOUserCell"
        let cell = (tableView.dequeueReusableCell(withIdentifier: reuseIdentifier) as? BOUserCell)
            ?? BOUserCell(style: .value1, reuseIdentifier: reuseIdentifier)

        if let user = users[indexPath.row] as? BOUser {
            cell.userIdLabel?.text = user.userId
            cell.scoreLabel?.text = "\(user.score)"
        }
        return cell
    }

    override func tableView(
        _ tableView: UITableView,
        commit editingStyle: UITableViewCell.EditingStyle,
        forRowAt indexPath: IndexPath
    ) {
        guard editingStyle == .delete else { return }

        let visibleUsers = NSMutableArray(array: users)
        visibleUsers.removeObject(at: indexPath.row)
        users = visibleUsers

        let defaults = UserDefaults.standard
        let noOfUsers = defaults.integer(forKey: "noOfUsers")
        var currentPlayer = defaults.integer(forKey: "currentPlayer")
        var tempUsers = defaults.array(forKey: "users") ?? []
        var tempDiffs = defaults.array(forKey: "diffs") ?? []
        var tempScores = defaults.array(forKey: "scores") ?? []
        var tempColors = defaults.array(forKey: "colors") ?? []

        tempUsers.remove(at: indexPath.row)
        tempDiffs.remove(at: indexPath.row)
        tempScores.remove(at: indexPath.row)
        tempColors.remove(at: indexPath.row)

        defaults.set(tempUsers, forKey: "users")
        defaults.set(tempDiffs, forKey: "diffs")
        defaults.set(tempScores, forKey: "scores")
        defaults.set(tempColors, forKey: "colors")
        defaults.set(noOfUsers - 1, forKey: "noOfUsers")

        if currentPlayer == indexPath.row {
            currentPlayer = 0
        } else if currentPlayer > indexPath.row {
            currentPlayer -= 1
        }
        defaults.set(currentPlayer, forKey: "currentPlayer")
        defaults.synchronize()

        tableView.deleteRows(at: [indexPath], with: .automatic)
    }

    func boNewUserViewController(
        _ sender: BONewUserViewController,
        gotUserId uid: String,
        gotDifficulty diff: NSNumber,
        score: NSNumber,
        andGotBackgroundColor bkg: NSNumber
    ) {
        let newUser = BOUser(id: uid, difficulty: diff, score: score, backgroundColor: bkg)

        if newUser.userId.isEmpty {
            dismiss(animated: true)
            tableView.reloadData()
            return
        }

        let defaults = UserDefaults.standard
        let newUsers = (defaults.array(forKey: "users") as? [String] ?? []) + [uid]
        let newDiffs = (defaults.array(forKey: "diffs") as? [NSNumber] ?? []) + [diff]
        let newScores = (defaults.array(forKey: "scores") as? [NSNumber] ?? []) + [NSNumber(value: 0)]
        let newColors = (defaults.array(forKey: "colors") as? [NSNumber] ?? []) + [bkg]
        let noOfUsers = defaults.integer(forKey: "noOfUsers") + 1

        let appSettings: [String: Any] = [
            "currentPlayer": noOfUsers - 1,
            "noOfUsers": noOfUsers,
            "users": newUsers,
            "diffs": newDiffs,
            "scores": newScores,
            "colors": newColors
        ]

        defaults.register(defaults: appSettings)
        defaults.synchronize()

        let userArray = NSMutableArray()
        for index in 0 ..< newUsers.count {
            userArray.add(
                BOUser(
                    id: newUsers[index],
                    difficulty: newDiffs[index],
                    score: newScores[index],
                    backgroundColor: newColors[index]
                )
            )
        }
        users = userArray

        dismiss(animated: true)
        tableView.reloadData()
    }

    func newScore(_ sender: BOViewController, withScore myScore: Int32) {
        users = sender.users
        NSLog("Score=%d", myScore)
        dismiss(animated: true)
        tableView.reloadData()
    }
}
